tests/modules/blockedclient.c

```diff
@@ -46,15 +46,15 @@ void *worker(void *arg) {
     // Wait for thread
     pthread_join(tid, NULL);
 
-    // Release GIL
-    RedisModule_ThreadSafeContextUnlock(ctx);
-
     // Reply to client
     RedisModule_ReplyWithSimpleString(ctx, "OK");
 
     // Unblock client
     RedisModule_UnblockClient(bc, NULL);
 
+    // Release GIL
+    RedisModule_ThreadSafeContextUnlock(ctx);
+
```

Comment:
https://github.com/redis/redis/pull/12817/commits/6e4de39a94bd782959d46da2e5bd665a68e97535
I've made the `moduleBlockedClientTimedOut()` distinguish between the main thread andmodule threads.
Since we have no way of knowing if `RM_UnblockClient()` was called in the thread, we have to make it always asynchronous to handle unblock status updates.
And we need to make the ctx with REDISMODULE_CTX_THREAD_SAFE flag, otherwise replying error in the `timeout_callback()` will not be thread-safe.

Reply:
looking at your changes in moduleBlockedClientTimedOut, i'm not sure i understand why we don't need to call updateStatsOnUnblock when the code comes from RM_UnblockClient.

Reply:
please see https://github.com/redis/redis/pull/12817#discussion_r1419883004